### ISSUE TITLE
Question about TF multithread session run with horovod

### ISSUE THREAD

**Issue by iamyangchen** [horovod/horovod#1862]
I'm trying to adopt horovod to scale our current TF training script to multiple gpus/machines. I'm testing with one machine with 4 GPUs right now, using command like `horovodrun -np 4 python trainer.py ...`

Our current script used python multithreading to speed up the training throughput, which basically calls session.run on each of the thread using the same session object. It seems that it's not possible to do this when adopting horovod, as it gives me error like "Invalid argument: Requested to allreduce, allgather, or broadcast a tensor with the same name as another tensor that is currently being processed." when I launch multiple threads. But it works if I limit the thread to only one. 

So I'm wondering if this is expected or is some configuration I have missed?

**Comment by tgaddair** [comment]
Hey @iamyangchen, it sounds like your multithreading approach is doing something similar to Horovod but within a single GPU: processing multiple batches in parallel.  Are you also averaging the gradients among the different threads or just applying them as soon as they're computed.  If you're not averaging them, I would expect you to have convergence problems.

If I understand what you're doing correctly, then I would suggest that spawning multiple threads within a Horovod worker / GPU should not buy you anything.  If you're getting a performance benefit from processing multiple batches in parallel on a single GPU, then it likely means that you're not fully utilizing the bandwidth of your GPU each batch.  A better solution would be to simply increase the batch size.

If you still believe it makes sense to process multiple batches in parallel on a single GPU, then my suggestion would be to replace the python multithreading with spawning more Horovod workers, e.g.:

```
horovodrun -np 8 python trainer.py
```

Then, to allow multiple processes to use the same GPU, you can do something like this:

```
config.gpu_options.visible_device_list = str(hvd.local_rank() % hvd.local_size())
```

But again, I think your best bet is to try increasing the batch size per GPU instead of attempting to process multiple batches on a single GPU at once.

**Comment by iamyangchen** [comment]
@tgaddair Thanks for the quick response and clarification!

I have actually tried use more horovod workers, basically do what you suggested. But then I end up with a NCCL error 

```
[1,4]<stderr>:: ncclCommInitRank failed: invalid usage
[1,4]<stderr>:	 [[node HorovodBroadcast_optimizer_xyz_bias_adam_amsgrad_optimizer_1_0 (defined at /usr/local/lib/python2.7/dist-packages/tensorflow_core/python/framework/ops.py:1748) ]]
```

adding `--mpi-args="-x NCCL_DEBUG=WARN"` to the command I saw warnings like
```
init.cc:598 NCCL WARN Duplicate GPU detected : rank 4 and rank 0 both on CUDA device 1b0
```

so I suspect NCCL only allows one process to pin to one GPU? Any suggestions on this error? 

BTW I'll give it a shot on the larger batch size.

**Comment by tgaddair** [comment]
@iamyangchen yes, it seems NCCL does not support using the same device for multiple ranks (see [docs](https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/usage/communicators.html)).

I'm sure we could think of a way to work around this, but I would still suggest trying to tune the batch size first.  In nearly every case, you should be able to saturate the GPU effectively with a single thread.
